ray-curator/ray_curator/stages/deduplication/fuzzy/minhash.py

```diff
@@ -0,0 +1,326 @@
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, Any, Literal
+
+import cudf
+import numpy as np
+import rmm
+
+from ray_curator.stages.base import ProcessingStage
+from ray_curator.stages.deduplication.id_generator import (
+    CURATOR_DEDUP_ID_STR,
+    CURATOR_ID_GENERATOR_ACTOR_NAME,
+    IdGenerator,
+)
+from ray_curator.stages.deduplication.io_utils import DeduplicationIO
+from ray_curator.stages.resources import Resources
+from ray_curator.tasks import FileGroupTask
+from ray_curator.utils.file_utils import get_fs
+
+if TYPE_CHECKING:
+    from ray_curator.backends.base import WorkerMetadata
+
+
+class MinHash(ABC):
+    """
+    Base class for computing minhash signatures of a document corpus
+    """
+
+    def __init__(
+        self,
+        seed: int = 42,
+        num_hashes: int = 260,
+        char_ngrams: int = 24,
+        use_64bit_hash: bool = False,
+    ):
+        """
+        Parameters
+        ----------
+        seed: Seed for minhash permutations
+        num_hashes: Length of minhash signature (No. of minhash permutations)
+        char_ngrams: Width of text window (in characters) while computing minhashes.
+        use_64bit_hash: Whether to use a 64 bit hash function.
+        """
+        self.num_hashes = num_hashes
+        self.char_ngram = char_ngrams
+        self.seed = seed
+        self.use_64bit_hash = use_64bit_hash
+
+    def generate_seeds(self, n_permutations: int = 260, seed: int = 0, bit_width: int = 32) -> np.ndarray:
+        """
+        Generate seeds for all minhash permutations based on the given seed.
+        This is a placeholder that child classes should implement if needed.
+        """
+        msg = "Child classes should implement this method if needed"
+        raise NotImplementedError(msg)
+
+    @abstractmethod
+    def compute_minhashes(self, text_series: Any) -> Any:  # noqa: ANN401
+        """
+        Compute minhash signatures for the given dataframe text column.
+        """
+
+
+class GPUMinHash(MinHash):
+    def __init__(
+        self,
+        seed: int = 42,
+        num_hashes: int = 260,
+        char_ngrams: int = 24,
+        use_64bit_hash: bool = False,
+        pool: bool = False,
+    ):
+        # Initialize parent class
+        MinHash.__init__(
+            self,
+            seed=seed,
+            num_hashes=num_hashes,
+            char_ngrams=char_ngrams,
+            use_64bit_hash=use_64bit_hash,
+        )
+
+        # Initialize memory pool for cuDF
+        if pool:
+            rmm.reinitialize(pool_allocator=pool)
```

Comment:
Curious about what is happening here. All the tests are `pool=False` while the default in the minhash stage is `pool=True`. Do we still need to worry about memory pools if we are controlling 1 GPU for this stage and each `GPUMinHashStage` actor is independent of the others?